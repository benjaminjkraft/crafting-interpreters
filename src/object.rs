use std::fmt;

#[derive(Debug)]
pub enum Object {
    Int(i64),
    Float(f64),
    Bool(bool),
    String(String),
    Nil,
}

impl fmt::Display for Object {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Object::Int(v) => write!(f, "{}", v),
            Object::Float(v) => write!(f, "{}", v),
            Object::Bool(v) => write!(f, "{}", v),
            Object::String(v) => write!(f, "{}", v),
            Object::Nil => write!(f, "nil"),
        }
    }
}
