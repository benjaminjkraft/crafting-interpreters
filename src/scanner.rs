use crate::error::LoxError;
use once_cell::sync::Lazy;
use std::collections::HashMap;
use std::fmt;
use TokenType::*;

pub struct Scanner<'src> {
    source: &'src str,
    start: usize,
    current: usize,
    line: usize,
}

impl<'src> Scanner<'src> {
    fn scan_tokens(&mut self) -> Result<Vec<Token<'src>>, LoxError> {
        let mut tokens = Vec::new();
        while !self.is_at_end() {
            self.start = self.current;
            if let Some(tok) = self.scan_token()? {
                tokens.push(tok);
            }
        }

        tokens.push(Token {
            type_: EOF,
            lexeme: "",
            line: self.line,
        });
        Ok(tokens)
    }

    fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }

    fn scan_token(&mut self) -> Result<Option<Token<'src>>, LoxError> {
        let c = self.advance();
        Ok(match c {
            b'(' => self.token(LeftParen),
            b')' => self.token(RightParen),
            b'{' => self.token(LeftBrace),
            b'}' => self.token(RightBrace),
            b',' => self.token(Comma),
            b'.' => self.token(Dot),
            b'-' => self.token(Minus),
            b'+' => self.token(Plus),
            b';' => self.token(Semicolon),
            b'*' => self.token(Star),
            b'!' => {
                let next_eq = self.match_(b'=');
                self.token(if next_eq { BangEqual } else { Bang })
            }
            b'=' => {
                let next_eq = self.match_(b'=');
                self.token(if next_eq { EqualEqual } else { Equal })
            }
            b'<' => {
                let next_eq = self.match_(b'=');
                self.token(if next_eq { LessEqual } else { Less })
            }
            b'>' => {
                let next_eq = self.match_(b'=');
                self.token(if next_eq { GreaterEqual } else { Greater })
            }
            b'/' => {
                if self.match_(b'/') {
                    self.advance_all(|c| c != b'\n');
                    None
                } else {
                    self.token(Slash)
                }
            }
            b' ' | b'\r' | b'\t' => None,
            b'\n' => None,
            b'"' => self.string()?,
            c => {
                if is_digit(c) {
                    self.number()
                } else if is_alpha(c) {
                    self.identifier()
                } else {
                    self.err(format!("Unexpected character: '{}'.", c))?
                }
            }
        })
    }

    fn err(&mut self, message: String) -> Result<Option<Token<'src>>, LoxError> {
        Err(LoxError {
            line: self.line,
            loc: String::new(),
            exit: 65,
            message,
        })
    }

    fn peek(&mut self) -> Option<u8> {
        if self.is_at_end() {
            None
        } else {
            Some(self.source.as_bytes()[self.current])
        }
    }

    fn peek_next(&mut self) -> Option<u8> {
        if self.current + 1 > self.source.len() {
            None
        } else {
            Some(self.source.as_bytes()[self.current + 1])
        }
    }

    fn advance(&mut self) -> u8 {
        let ch = self.peek().unwrap();
        if ch == b'\n' {
            self.line += 1;
        }
        self.current += 1;
        ch
    }

    fn match_(&mut self, expected: u8) -> bool {
        self.match_pred(&|c| c == expected)
    }

    fn advance_all(&mut self, pred: impl Fn(u8) -> bool) {
        while self.match_pred(&pred) {}
    }

    fn match_pred(&mut self, pred: &impl Fn(u8) -> bool) -> bool {
        if self.peek().is_some_and(pred) {
            self.advance();
            true
        } else {
            false
        }
    }

    fn token(&self, type_: TokenType) -> Option<Token<'src>> {
        Some(Token {
            type_,
            lexeme: &self.source[self.start..self.current],
            line: self.line,
        })
    }

    fn string(&mut self) -> Result<Option<Token<'src>>, LoxError> {
        self.advance_all(|c| c != b'"');
        if !self.match_(b'"') {
            self.err("Unterminated string".to_string())?;
        }

        Ok(self.token(StringLiteral))
    }

    fn number(&mut self) -> Option<Token<'src>> {
        self.advance_all(is_digit);
        let decimal = self.peek() == Some(b'.') && self.peek_next().is_some_and(is_digit);
        if decimal {
            self.advance();
            self.advance_all(is_digit);
        }
        self.token(Number)
    }

    fn identifier(&mut self) -> Option<Token<'src>> {
        self.advance_all(is_alpha_numeric);

        let type_ = KEYWORDS
            .get(&self.source[self.start..self.current])
            .copied()
            .unwrap_or(Identifier);
        self.token(type_)
    }
}

fn is_digit(c: u8) -> bool {
    c >= b'0' && c <= b'9'
}

fn is_alpha(c: u8) -> bool {
    c >= b'a' && c <= b'z' || c >= b'A' && c <= b'Z' || c == b'_'
}

fn is_alpha_numeric(c: u8) -> bool {
    is_digit(c) || is_alpha(c)
}

pub fn scan_tokens<'src>(source: &'src str) -> Result<Vec<Token<'src>>, LoxError> {
    let mut scanner = Scanner {
        source,
        start: 0,
        current: 0,
        line: 1,
    };
    scanner.scan_tokens()
}

#[derive(Debug, Clone)]
pub struct Token<'src> {
    pub type_: TokenType,
    pub lexeme: &'src str,
    pub line: usize,
}

impl<'src> fmt::Display for Token<'src> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?} {}", self.type_, self.lexeme)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenType {
    // Single-character tokens.
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,

    // One or two character tokens.
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,

    // Literals.
    Identifier,
    StringLiteral,
    Number,

    // Keywords.
    And,
    Class,
    Else,
    False,
    Fun,
    For,
    If,
    Nil,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,

    EOF,
}

static KEYWORDS: Lazy<HashMap<&str, TokenType>> = Lazy::new(|| {
    let mut m = HashMap::new();
    m.insert("and", And);
    m.insert("class", Class);
    m.insert("else", Else);
    m.insert("false", False);
    m.insert("for", For);
    m.insert("fun", Fun);
    m.insert("if", If);
    m.insert("nil", Nil);
    m.insert("or", Or);
    m.insert("print", Print);
    m.insert("return", Return);
    m.insert("super", Super);
    m.insert("this", This);
    m.insert("true", True);
    m.insert("var", Var);
    m.insert("while", While);
    m
});

#[test]
fn test_scanner() {
    insta::assert_debug_snapshot!(scan_tokens("(){},.-+;* // (symbols)"));
    insta::assert_debug_snapshot!(scan_tokens(""));
    insta::assert_debug_snapshot!(scan_tokens("\n\n"));
    insta::assert_debug_snapshot!(scan_tokens("// asdf"));
    insta::assert_debug_snapshot!(scan_tokens("// asdf\n"));
    insta::assert_debug_snapshot!(scan_tokens("\n\n!a and !!b and !!c != d == e////"));
    insta::assert_debug_snapshot!(scan_tokens("= == < <= > >= => =<"));
    insta::assert_debug_snapshot!(scan_tokens("1/1.1/1.23/123.45"));
    insta::assert_debug_snapshot!(scan_tokens("1"));
    insta::assert_debug_snapshot!(scan_tokens("\n\n\t\t   \n\t\t   \"asdf!!\"\n\nvar2"));
    insta::assert_debug_snapshot!(scan_tokens("and class class_ else false for fun"));
    insta::assert_debug_snapshot!(scan_tokens("if if_ nil null or print return super"));
    insta::assert_debug_snapshot!(scan_tokens("this true var while class and fun"));
}
