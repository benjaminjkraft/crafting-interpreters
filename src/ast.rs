use crate::object;
use crate::scanner;
use derive_more::From;

#[derive(Debug)]
pub struct Program<'src> {
    pub stmts: Vec<Stmt<'src>>,
}

#[derive(Debug, From)]
pub enum Expr<'src> {
    Assign(AssignExpr<'src>),
    Binary(BinaryExpr<'src>),
    Call(CallExpr<'src>),
    Grouping(GroupingExpr<'src>),
    Literal(LiteralExpr),
    Logical(LogicalExpr<'src>),
    Unary(UnaryExpr<'src>),
    Variable(VariableExpr<'src>),
}

#[derive(Debug, From)]
pub enum Stmt<'src> {
    Block(BlockStmt<'src>),
    Expr(ExprStmt<'src>),
    Function(FunctionStmt<'src>),
    If(IfStmt<'src>),
    Print(PrintStmt<'src>),
    Var(VarStmt<'src>),
    While(WhileStmt<'src>),
}

#[derive(Debug)]
pub struct AssignExpr<'src> {
    pub name: scanner::Token<'src>,
    pub value: Box<Expr<'src>>,
}

#[derive(Debug)]
pub struct BinaryExpr<'src> {
    pub left: Box<Expr<'src>>,
    pub operator: scanner::Token<'src>,
    pub right: Box<Expr<'src>>,
}

#[derive(Debug)]
pub struct CallExpr<'src> {
    pub callee: Box<Expr<'src>>,
    pub paren: scanner::Token<'src>,
    pub arguments: Vec<Expr<'src>>,
}

#[derive(Debug)]
pub struct GroupingExpr<'src> {
    pub expr: Box<Expr<'src>>,
}

#[derive(Debug)]
pub struct LiteralExpr {
    pub value: object::Literal,
}

#[derive(Debug)]
pub struct LogicalExpr<'src> {
    pub left: Box<Expr<'src>>,
    pub operator: scanner::Token<'src>,
    pub right: Box<Expr<'src>>,
}

#[derive(Debug)]
pub struct UnaryExpr<'src> {
    pub operator: scanner::Token<'src>,
    pub right: Box<Expr<'src>>,
}

#[derive(Debug)]
pub struct VariableExpr<'src> {
    pub name: scanner::Token<'src>,
}

#[derive(Debug)]
pub struct BlockStmt<'src> {
    pub stmts: Vec<Stmt<'src>>,
}

#[derive(Debug)]
pub struct ExprStmt<'src> {
    pub expr: Box<Expr<'src>>,
}

#[derive(Debug)]
pub struct FunctionStmt<'src> {
    pub name: scanner::Token<'src>,
    pub parameters: Vec<scanner::Token<'src>>,
    pub body: Vec<Stmt<'src>>,
}

#[derive(Debug)]
pub struct IfStmt<'src> {
    pub condition: Box<Expr<'src>>,
    pub then_: Box<Stmt<'src>>,
    pub else_: Option<Box<Stmt<'src>>>,
}

#[derive(Debug)]
pub struct PrintStmt<'src> {
    pub expr: Box<Expr<'src>>,
}

#[derive(Debug)]
pub struct VarStmt<'src> {
    pub name: scanner::Token<'src>,
    pub initializer: Option<Box<Expr<'src>>>,
}

#[derive(Debug)]
pub struct WhileStmt<'src> {
    pub condition: Box<Expr<'src>>,
    pub body: Box<Stmt<'src>>,
}
