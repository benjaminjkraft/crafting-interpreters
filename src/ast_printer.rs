use crate::ast::Visited;
use crate::ast::*;
use crate::object;
use crate::scanner;
use itertools::Itertools;

pub struct AstPrinter {}

impl<'a> AstPrinter {
    pub fn print(&self, expr: Expr<'a>) -> String {
        expr.accept(self)
    }

    fn parenthesize(&self, name: &'a str, exprs: Vec<&Box<Expr<'a>>>) -> String {
        return format!(
            "({}{})",
            name,
            exprs
                .into_iter()
                .map(|e| format!(" {}", (*e).accept(self)))
                .join("")
        );
    }
}

impl<'a> Visitor<'a, String> for &AstPrinter {
    fn visit_binary_expr(&self, node: &BinaryExpr<'a>) -> String {
        self.parenthesize(node.operator.lexeme, vec![&node.left, &node.right])
    }
    fn visit_grouping_expr(&self, node: &GroupingExpr<'a>) -> String {
        self.parenthesize("group", vec![&node.expr])
    }
    fn visit_literal_expr(&self, node: &LiteralExpr) -> String {
        self.parenthesize(&node.value.to_string(), vec![])
    }
    fn visit_unary_expr(&self, node: &UnaryExpr<'a>) -> String {
        self.parenthesize(node.operator.lexeme, vec![&node.right])
    }
}

#[test]
fn test_printer() {
    let expr = Expr::Binary(BinaryExpr {
        left: Box::new(Expr::Unary(UnaryExpr {
            operator: scanner::Token {
                type_: scanner::TokenType::Minus,
                lexeme: "-",
                literal: object::Object::Nil,
                line: 1,
            },
            right: Box::new(Expr::Literal(LiteralExpr {
                value: object::Object::Int(123),
            })),
        })),
        operator: scanner::Token {
            type_: scanner::TokenType::Star,
            lexeme: "*",
            literal: object::Object::Nil,
            line: 1,
        },
        right: Box::new(Expr::Grouping(GroupingExpr {
            expr: Box::new(Expr::Literal(LiteralExpr {
                value: object::Object::Float(45.67),
            })),
        })),
    });
    let printer = AstPrinter {};
    insta::assert_debug_snapshot!(printer.print(expr))
}
