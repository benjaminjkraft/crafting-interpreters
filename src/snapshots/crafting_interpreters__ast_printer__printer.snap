---
source: src/ast_printer.rs
expression: printer.print(expr)
---
"(* (- (123)) (group (45.67)))"
